command/agent/agent.go

```diff
@@ -518,6 +528,66 @@ func (a *Agent) setupServer() error {
 	return nil
 }
 
+// setupNodeID will pull the persisted node ID, if any, or create a random one
+// and persist it.
+func (a *Agent) setupNodeID(config *nomad.Config) error {
+	// For dev mode we have no filesystem access so just make a node ID.
+	if a.config.DevMode {
+		config.NodeID = uuid.Generate()
+		return nil
+	}
+
+	// Load saved state, if any. Since a user could edit this, we also
+	// validate it. Saved state overwrites any configured node id
+	fileID := filepath.Join(config.DataDir, "node-id")
```

Comment:
Should we reuse this method for the client which currently stores its ID in `client/client-id`? https://github.com/hashicorp/nomad/blob/98a403a5a692efc21a608167656e47de9f0934fe/client/client.go#L837-L842

Even if we want to share the code it's not something we need to do in this PR (or even in 0.8).

Reply:
Consul uses `${data-dir}/node-id` would be +1 on being consistent on that :)